File: 02-dive/solution.js
const input = require(`fs`)
  .readFileSync(0)
  .toString()
  .trim()

const part1 = () => {
  let forward = 0, down = 0, up = 0
  eval(input.replace(/ /g, `+=`))
  console.log(forward * (down - up))
}

part1()

const part2 = () => {
  let forward = 0, aim = 0, depth = 0
  eval(
    input
      .replace(/forward (\d+)/g, `forward += $1; depth += aim * $1;`)
      .replace(/up (\d+)/g, `aim -= $1`)
      .replace(/down (\d+)/g, `aim += $1`)
  )

  console.log(forward * depth)
}

part2()

File: 02-dive/solution.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("./input")
        .expect("To read the file");
    let (x, y, aim, y2) = input.lines()
        .map(|x| match x.split(" ").collect::<Vec<_>>()[..] {
            ["forward", d] => (d.parse::<i32>().unwrap(), 0),
            ["down", d] => (0, d.parse::<i32>().unwrap()),
            ["up", d] => (0, -d.parse::<i32>().unwrap()),
            _ => (0, 0)
        })
        .fold(
            (0, 0, 0, 0),
            |(x, y1, aim, y2), (dx, dy)| (x + dx, y1 + dy, aim + dy, y2 + dx * aim)
        );

    println!("part 1: {}, aim: {}, part 2: {}", x * y, aim, x * y2)
}
